### PR TITLE
Optimize client entry module by tracking individual imported names from the client boundary

### PR DESCRIPTION
**Pull request by shuding**
Previously, we're collecting all client module imports from the server layer and then create the client entry module with these imports (`import(/* webpackMode: "eager" */ importPath)`). Since that's a separate module graph, Webpack will just keep everything in the bundle.

That means if you only import one value from it (`import { Foo } from './client-module'`), the entire `./client-module` will be included in the client bundle.

This PR adds the tracking for actual imported names too, and create the client entry imports with the `/* webpackExports */` comment so Webpack knows which exports are needed and can tree-shake the unused ones. In certain cases this can reduce the size of the client bundle quite a lot.

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 29000b53812d93d1523b83394e893083c15318dc

`pnpm test-start test/e2e/app-dir/app-edge/app-edge.test.ts` 
- app-dir edge SSR > should handle edge only routes
- app-dir edge SSR > should retrieve cookies in a server component in the edge runtime
- app-dir edge SSR > should handle /index routes correctly

<details>
<summary>Expand output</summary>

  ● app-dir edge SSR › should handle edge only routes

    expect(received).toContain(expected) // indexOf

    Expected substring: "<p>Edge!</p>"
    Received string:    "Internal Server Error"

      11 |     it('should handle edge only routes', async () => {
      12 |       const appHtml = await next.render('/edge/basic')
    > 13 |       expect(appHtml).toContain('<p>Edge!</p>')
         |                       ^
      14 |
      15 |       const pageHtml = await next.render('/pages-edge')
      16 |       expect(pageHtml).toContain('<p>pages-edge-ssr</p>')

      at Object.toContain (e2e/app-dir/app-edge/app-edge.test.ts:13:23)

  ● app-dir edge SSR › should retrieve cookies in a server component in the edge runtime

    expect(received).toInclude(expected)

    Expected string to include:
      "Hello"
    Received:
      "Internal Server Error"

      19 |     it('should retrieve cookies in a server component in the edge runtime', async () => {
      20 |       const res = await next.fetch('/edge-apis/cookies')
    > 21 |       expect(await res.text()).toInclude('Hello')
         |                                ^
      22 |     })
      23 |
      24 |     it('should handle /index routes correctly', async () => {

      at Object.toInclude (e2e/app-dir/app-edge/app-edge.test.ts:21:32)

  ● app-dir edge SSR › should handle /index routes correctly

    expect(received).toContain(expected) // indexOf

    Expected substring: "the /index route"
    Received string:    "Internal Server Error"

      24 |     it('should handle /index routes correctly', async () => {
      25 |       const appHtml = await next.render('/index')
    > 26 |       expect(appHtml).toContain('the /index route')
         |                       ^
      27 |     })
      28 |
      29 |     if ((globalThis as any).isNextDev) {

      at Object.toContain (e2e/app-dir/app-edge/app-edge.test.ts:26:23)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| buildDuration | 10.3s | 10.5s | ⚠️ +117ms |
| buildDurationCached | 6s | 5.9s | N/A |
| nodeModulesSize | 174 MB | 174 MB | ⚠️ +7.43 kB |
| nextStartRea..uration (ms) | 526ms | 511ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| 199-HASH.js gzip | 27.5 kB | 27.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.3 kB | 45.3 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 250 B | N/A |
| main-HASH.js gzip | 32.9 kB | 32.9 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.8 kB | 98.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | N/A |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.36 kB | 4.36 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.63 kB | N/A |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.08 kB | 1.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 543 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 524 B | 524 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.7 kB | ✓ |
| page.js gzip | 155 kB | 165 kB | ⚠️ +9.49 kB |
| Overall change | 249 kB | 258 kB | ⚠️ +9.49 kB |

</details>

<details>
<summary><strong>Middleware size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js shu/608b | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 623 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 22.5 kB | 22.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 2.3 kB | ⚠️ +373 B |
| Overall change | 1.92 kB | 2.3 kB | ⚠️ +373 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-runtime-webpack.js</strong></summary>

```diff
@@ -111,6 +111,54 @@
     /******/
   })();
   /******/
+  /******/ /* webpack/runtime/create fake namespace object */
+  /******/ (() => {
+    /******/ var getProto = Object.getPrototypeOf
+      ? (obj) => Object.getPrototypeOf(obj)
+      : (obj) => obj.__proto__;
+    /******/ var leafPrototypes;
+    /******/ // create a fake namespace object
+    /******/ // mode & 1: value is a module id, require it
+    /******/ // mode & 2: merge all properties of value into the ns
+    /******/ // mode & 4: return value when already ns object
+    /******/ // mode & 16: return value when it's Promise-like
+    /******/ // mode & 8|1: behave like require
+    /******/ __webpack_require__.t = function (value, mode) {
+      /******/ if (mode & 1) value = this(value);
+      /******/ if (mode & 8) return value;
+      /******/ if (typeof value === "object" && value) {
+        /******/ if (mode & 4 && value.__esModule) return value;
+        /******/ if (mode & 16 && typeof value.then === "function")
+          return value;
+        /******/
+      }
+      /******/ var ns = Object.create(null);
+      /******/ __webpack_require__.r(ns);
+      /******/ var def = {};
+      /******/ leafPrototypes = leafPrototypes || [
+        null,
+        getProto({}),
+        getProto([]),
+        getProto(getProto),
+      ];
+      /******/ for (
+        var current = mode & 2 && value;
+        typeof current == "object" && !~leafPrototypes.indexOf(current);
+        current = getProto(current)
+      ) {
+        /******/ Object.getOwnPropertyNames(current).forEach(
+          (key) => (def[key] = () => value[key])
+        );
+        /******/
+      }
+      /******/ def["default"] = () => value;
+      /******/ __webpack_require__.d(ns, def);
+      /******/ return ns;
+      /******/
+    };
+    /******/
+  })();
+  /******/
   /******/ /* webpack/runtime/define property getters */
   /******/ (() => {
     /******/ // define getter functions for harmony exports
```
</details>
<details>
<summary>Diff for <strong>199-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 29000b53812d93d1523b83394e893083c15318dc

File: packages/next/src/build/webpack/loaders/next-flight-client-entry-loader.ts [hunk 1]
```diff
@@ -15,19 +15,21 @@ export default function transformSource(this: any) {
...
-        `import(/* webpackMode: "eager" */ ${JSON.stringify(request)})`
-    )
+    .filter(([request]) => (isServer ? !regexCSS.test(request) : true))
+    .map(([request, imports]) => {
+      if (imports[0] === '*') {
+        return `import(/* webpackMode: "eager" */ ${JSON.stringify(request)})`
+      }
+
+      const uniqueImports = [...new Set(imports)]
+      return `import(/* webpackMode: "eager" *//* webpackExports: [${uniqueImports
```
Comment (sokra) [inline packages/next/src/build/webpack/loaders/next-flight-client-entry-loader.ts:29]:
```suggestion
      return `import(/* webpackMode: "eager", webpackExports: [${uniqueImports
```

a comma would do it too

File: packages/next/src/build/webpack/loaders/next-flight-client-entry-loader.ts [hunk 2]
```diff
@@ -15,19 +15,21 @@ export default function transformSource(this: any) {
...
+        .map((i) => JSON.stringify(i))
+        .join(', ')}] */ ${JSON.stringify(request)})`
```
Comment (sokra) [inline packages/next/src/build/webpack/loaders/next-flight-client-entry-loader.ts:31]:
```suggestion
      return `import(/* webpackMode: "eager", webpackExports: ${JSON.stringify(uniqueImports)} */ ${JSON.stringify(request)})`
```

**Pull request closed without merging**